File: frontend/src/lib/weapons.ts
// Ids must match the backend catalogue 

export interface WeaponOption {
  id: string
  label: string
}

export interface WeaponCategory {
  id: string
  weapons: WeaponOption[]
}

export const WEAPON_CATEGORIES: WeaponCategory[] = [
  {
    id: 'pistols',
    weapons: [
      { id: 'glock', label: 'Glock-18' },
      { id: 'usp_s', label: 'USP-S' },
      { id: 'p2000', label: 'P2000' },
      { id: 'p250', label: 'P250' },
      { id: 'five_seven', label: 'Five-SeveN' },
      { id: 'tec9', label: 'Tec-9' },
      { id: 'cz75', label: 'CZ75-Auto' },
      { id: 'deagle', label: 'Desert Eagle' },
      { id: 'dualies', label: 'Dual Berettas' },
      { id: 'revolver', label: 'R8 Revolver' },
    ],
  },
  {
    id: 'smgs',
    weapons: [
      { id: 'mac10', label: 'MAC-10' },
      { id: 'mp9', label: 'MP9' },
      { id: 'mp7', label: 'MP7' },
      { id: 'mp5', label: 'MP5-SD' },
      { id: 'ump45', label: 'UMP-45' },
      { id: 'p90', label: 'P90' },
      { id: 'bizon', label: 'PP-Bizon' },
    ],
  },
  {
    id: 'rifles',
    weapons: [
      { id: 'galil', label: 'Galil AR' },
      { id: 'famas', label: 'FAMAS' },
      { id: 'ak47', label: 'AK-47' },
      { id: 'm4a4', label: 'M4A4' },
      { id: 'm4a1_s', label: 'M4A1-S' },
      { id: 'sg553', label: 'SG 553' },
      { id: 'aug', label: 'AUG' },
    ],
  },
  {
    id: 'snipers',
    weapons: [
      { id: 'ssg08', label: 'SSG 08' },
      { id: 'awp', label: 'AWP' },
      { id: 'g3sg1', label: 'G3SG1' },
      { id: 'scar20', label: 'SCAR-20' },
    ],
  },
  {
    id: 'heavy',
    weapons: [
      { id: 'nova', label: 'Nova' },
      { id: 'xm1014', label: 'XM1014' },
      { id: 'mag7', label: 'MAG-7' },
      { id: 'sawedoff', label: 'Sawed-Off' },
      { id: 'm249', label: 'M249' },
      { id: 'negev', label: 'Negev' },
    ],
  },
]

export const WEAPON_IDS: string[] = WEAPON_CATEGORIES.flatMap((c) => c.weapons.map((w) => w.id))

export const WEAPON_LABELS: Record<string, string> = Object.fromEntries(
  WEAPON_CATEGORIES.flatMap((c) => c.weapons.map((w) => [w.id, w.label])),
)

File: backend/app/domain/weapons.py
from __future__ import annotations

from collections.abc import Iterable

# CS2 weapon catalogue
WEAPON_CATALOG: tuple[tuple[str, str, tuple[str, ...]], ...] = (
    # pistols
    ("glock", "pistols", ("glock",)),
    ("usp_s", "pistols", ("usp",)),
    ("p2000", "pistols", ("hkp2000", "p2000")),
    ("p250", "pistols", ("p250",)),
    ("five_seven", "pistols", ("five-seven", "fiveseven")),
    ("tec9", "pistols", ("tec-9", "tec9")),
    ("cz75", "pistols", ("cz75",)),
    ("deagle", "pistols", ("deagle", "desert eagle")),
    ("dualies", "pistols", ("dual berettas", "dualberettas", "elite")),
    ("revolver", "pistols", ("revolver",)),
    # smgs
    ("mac10", "smgs", ("mac-10", "mac10")),
    ("mp9", "smgs", ("mp9",)),
    ("mp7", "smgs", ("mp7",)),
    ("mp5", "smgs", ("mp5",)),
    ("ump45", "smgs", ("ump",)),
    ("p90", "smgs", ("p90",)),
    ("bizon", "smgs", ("bizon",)),
    # rifles
    ("galil", "rifles", ("galil",)),
    ("famas", "rifles", ("famas",)),
    ("ak47", "rifles", ("ak-47", "ak47")),
    ("m4a4", "rifles", ("m4a4",)),
    ("m4a1_s", "rifles", ("m4a1-s", "m4a1_s", "m4a1")),
    ("sg553", "rifles", ("sg553", "sg 553", "sg556")),
    ("aug", "rifles", ("aug",)),
    # snipers
    ("ssg08", "snipers", ("ssg08", "ssg 08", "scout")),
    ("awp", "snipers", ("awp",)),
    ("g3sg1", "snipers", ("g3sg1",)),
    ("scar20", "snipers", ("scar-20", "scar20")),
    # heavy
    ("nova", "heavy", ("nova",)),
    ("xm1014", "heavy", ("xm1014",)),
    ("mag7", "heavy", ("mag-7", "mag7")),
    ("sawedoff", "heavy", ("sawed-off", "sawedoff", "sawed off")),
    ("m249", "heavy", ("m249",)),
    ("negev", "heavy", ("negev",)),
)

WEAPON_IDS: tuple[str, ...] = tuple(w[0] for w in WEAPON_CATALOG)


def weapons_present(inventory: Iterable[str]) -> list[str]:
    """Catalogue ids whose needle appears in any of the inventory strings."""
    hay = " ".join(str(x).lower() for x in inventory if x)
    return [wid for wid, _cat, needles in WEAPON_CATALOG if any(n in hay for n in needles)]
